logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb

```diff
@@ -6,130 +6,234 @@
+      CONTROLLER_SEPARATOR_RE = ","
+
+      def initialize
+        @cgroup_file = Pathname.new("/proc/self/cgroup")
+        @cpuacct_dir = Pathname.new("/sys/fs/cgroup/cpuacct")
+        @cpu_dir = Pathname.new("/sys/fs/cgroup/cpu")
+        @cpu_path_override = Override.new("ls.cgroup.cpu.path.override")
+        @cpuacct_path_override = Override.new("ls.cgroup.cpuacct.path.override")
+        @controller_cpuacct = "cpuacct"
+        @controller_cpu = "cpu"
+      end
 
-    class << self
-      def are_cgroup_available?
-        [::File.exist?(PROC_SELF_CGROUP_FILE),
-         Dir.exist?(PROC_CGROUP_CPU_DIR),
-         Dir.exist?(PROC_CGROUP_CPUACCT_DIR)].all?
+      def cgroup_available?
+        # don't cache to ivar, in case the files are mounted after logstash starts??
+        @cgroup_file.exist? && @cpuacct_dir.exist? && @cpu_dir.exist?
       end
 
-      def control_groups
+      def controller_groups
         response = {}
-
-        read_proc_self_cgroup_lines.each do |line|
+        read_lines.each do |line|
           matches = CONTROL_GROUP_RE.match(line)
+          next if matches.nil?
           # multiples controls, same hierarchy
           controllers = matches[1].split(CONTROLLER_SEPARATOR_RE)
-          controllers.each_with_object(response) { |controller| response[controller] = matches[2] }
+          controllers.each_with_object(response) do |controller|
+            case controller
+            when @controller_cpu
+              response[controller] = CpuResource.new(@cpu_dir, @cpu_path_override, matches[2])
+            when @controller_cpuacct
+              response[controller] = CpuAcctResource.new(@cpuacct_dir, @cpuacct_path_override, matches[2])
+            else
+              response[controller] = UnhandledResource.new(Pathname.new("/sys/fs/cgroup").join(controller), NullOverride.new, matches[2])
+            end
+          end
         end
-
         response
       end
 
-      def read_first_line(path)
-        IO.readlines(path).first
+      def read_lines
+        IO.readlines(@cgroup_file)
       end
+    end
 
-      def cgroup_cpuacct_usage_nanos(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group, CPUACCT_USAGE_FILE)).to_i
+    module ControllerResource
+      attr_reader :base_path, :override, :offset_path
+      def initialize(base_path, override, offset_path)
+        @base_path = base_path
+        # override is needed here for the logging statements
+        # so we do the override here and not in the caller
+        @override = override
+        @offset_path = @override.override(offset_path)
       end
-
-      def cgroup_cpu_fs_period_micros(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group, CPU_FS_PERIOD_US_FILE)).to_i
+      def check_path(path, not_found_value)
+        if exist?(path)
+          yield
+        else
+          message = "File #{path} cannot be found, "
+          if override.nil?
```

Comment:
not consistent/compatible with `NullOverride` abstraction we use when creating `UnhandledResource`, which includes this module. From my understanding, we don't actually _use_ `UnhandledResource#check_path` anywhere currently, but this could be problematic if we do.

Reply:
We know from the code that only deal with `cpu` and `cpuacct` and this code has not changed in two years so I think the chances of expansion to include values from other controllers is less likely. However, if that change was required then its obvious that one would create a new controller and remove the UnhandledResource for that controller key.